src/object/cli_obj.c

```diff
@@ -1462,49 +1423,521 @@ shard_task_sched(tse_task_t *task, void *arg)
+	case DAOS_OBJ_RPC_PUNCH:
+	case DAOS_OBJ_RPC_PUNCH_DKEYS:
+	case DAOS_OBJ_RPC_PUNCH_AKEYS:
+		return &obj_auxi->p_args.pa_auxi;
+	default:
+		D_ERROR("bad opc %d.\n", obj_auxi->opc);
+		return NULL;
+	};
+}
+
+static int
+shard_io(tse_task_t *task, struct shard_auxi_args *shard_auxi)
+{
+	struct dc_object		*obj = shard_auxi->obj;
+	struct obj_auxi_args		*obj_auxi = shard_auxi->obj_auxi;
+	struct dc_obj_shard		*obj_shard;
+	struct obj_req_tgts		*req_tgts;
+	struct daos_obj_shard_tgt	*fw_shard_tgts;
+	uint32_t			 fw_cnt;
+	bool				 is_shard_task;
+	int				 rc;
+
+	D_ASSERT(obj != NULL);
+	is_shard_task = !d_list_empty(&obj_auxi->shard_task_head);
+	rc = obj_shard_open(obj, shard_auxi->shard, shard_auxi->map_ver,
+			    &obj_shard);
+	if (rc != 0) {
+		/* skip a failed target */
+		if (rc == -DER_NONEXIST)
+			rc = 0;
+
+		if (is_shard_task)
+			tse_task_complete(task, rc);
+		return rc;
+	}
+
+	shard_auxi->flags = shard_auxi->obj_auxi->flags;
+	req_tgts = &shard_auxi->obj_auxi->req_tgts;
+	fw_shard_tgts = req_tgts->ort_srv_disp ?
+			req_tgts->ort_shard_tgts : NULL;
+	fw_cnt = req_tgts->ort_srv_disp ? req_tgts->ort_tgts_nr : 0;
+	rc = shard_auxi->shard_io_cb(obj_shard, obj_auxi->opc, shard_auxi,
+				     fw_shard_tgts, fw_cnt, task);
+	obj_shard_close(obj_shard);
+
+	return rc;
+}
+
+static int
+shard_io_task(tse_task_t *task)
+{
+	struct shard_auxi_args		*shard_auxi;
+
+	shard_auxi = tse_task_buf_embedded(task, sizeof(*shard_auxi));
+
+	return shard_io(task, shard_auxi);
+}
+
+typedef int (*shard_io_prep_cb_t)(tse_task_t *shard_task, shard_io_cb_t io_cb,
+				  struct dc_object *obj,
+				  struct obj_auxi_args *obj_auxi,
+				  struct daos_obj_shard_tgt *tgt,
+				  uint64_t dkey_hash, uint32_t map_ver,
+				  daos_epoch_t epoch);
+
+static int
+obj_req_fanout(struct dc_object *obj, struct obj_auxi_args *obj_auxi,
+	       uint64_t dkey_hash, uint32_t map_ver, daos_epoch_t epoch,
+	       shard_io_prep_cb_t io_prep_cb, shard_io_cb_t io_cb,
+	       tse_task_t *obj_task)
+{
+	struct obj_req_tgts		*req_tgts = &obj_auxi->req_tgts;
+	d_list_t			*task_list = &obj_auxi->shard_task_head;
+	tse_task_t			*shard_task;
+	struct daos_obj_shard_tgt	*tgt;
+	uint32_t			 i, tgts_nr;
+	int				 rc = 0;
+
+	/* for retried obj IO, reuse the previous shard tasks and resched it */
+	if (obj_auxi->io_retry) {
+		/* We mark the RPC as RESEND although @io_retry does not
+		 * guarantee that the RPC has ever been sent. It may cause
+		 * some overhead on server side, but no correctness issues.
+		 *
+		 * On the other hand, the client may resend the RPC to new
+		 * shard if leader switched. That is why the resend logic
+		 * is handled at object layer rather than shard layer.
+		 */
+		obj_auxi->flags = ORF_RESEND;
+
+		/* if with shard task list, just reuse it and re-schedule */
+		if (!d_list_empty(task_list))
+			goto task_sched;
+
+		/* for srv dispatch request, the task_list is empty, below code
+		 * will reset leader's shard and target.
```

Comment:
because when retry & srv_dispatch, the leader possibly changed to other shard right?
it is for the same purpose of your original shard_task_reset_target().

Reply:
Yes, I see.